### PR TITLE
Add ability to toggle all free mods quickly at multiplayer song select

### PR DESCRIPTION
**Pull request by peppy**
Addresses concerns in https://github.com/ppy/osu/discussions/23560.

https://github.com/ppy/osu/assets/191335/e91915b9-97fe-42b4-9061-814cf3f10ff6

### CODE REVIEW

**Comment by Haspamelodica** [comment]
Suggestion: Maybe change "Off" and "On" to "None" and "All"?
Off/On has the advantage of being very intuitive for new users, but None/All describes better what is actually happening. I'm not sure which is better here.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 1]
```diff
@@ -111,6 +111,10 @@ protected virtual IEnumerable<ShearedButton> CreateFooterButtons()
 
         private readonly Bindable<Dictionary<ModType, IReadOnlyList<Mod>>> globalAvailableMods = new Bindable<Dictionary<ModType, IReadOnlyList<Mod>>>();
 
+        public IEnumerable<Mod> AllAvailableAndValidMods => allAvailableMods
+                                                            .Select(s => s.Mod)
+                                                            .Where(m => isValidMod(m));
```
Comment (bdach) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs:116]: I don't really like that this is exposed. Feels much too specific.

I'd propose the following equivalent instead:

```diff
diff --git a/osu.Game/Overlays/Mods/ModSelectOverlay.cs b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
index ba3f01a688..7ec108e3ec 100644
--- a/osu.Game/Overlays/Mods/ModSelectOverlay.cs
+++ b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
@@ -111,11 +111,7 @@ protected virtual IEnumerable<ShearedButton> CreateFooterButtons()
 
         private readonly Bindable<Dictionary<ModType, IReadOnlyList<Mod>>> globalAvailableMods = new Bindable<Dictionary<ModType, IReadOnlyList<Mod>>>();
 
-        public IEnumerable<Mod> AllAvailableAndValidMods => allAvailableMods
-                                                            .Select(s => s.Mod)
-                                                            .Where(m => isValidMod(m));
-
-        private IEnumerable<ModState> allAvailableMods => AvailableMods.Value.SelectMany(pair => pair.Value);
+        public IEnumerable<ModState> AllAvailableMods => AvailableMods.Value.SelectMany(pair => pair.Value);
 
         private readonly BindableBool customisationVisible = new BindableBool();
 
@@ -386,7 +382,7 @@ private void createLocalMods()
 
         private void filterMods()
         {
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
                 modState.ValidForSelection.Value = modState.Mod.HasImplementation && IsValidMod.Invoke(modState.Mod);
         }
 
@@ -411,7 +407,7 @@ private void updateCustomisation()
             bool anyCustomisableModActive = false;
             bool anyModPendingConfiguration = false;
 
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
             {
                 anyCustomisableModActive |= modState.Active.Value && modState.Mod.GetSettingsSourceProperties().Any();
                 anyModPendingConfiguration |= modState.PendingConfiguration;
@@ -468,7 +464,7 @@ private void updateFromExternalSelection()
 
             var newSelection = new List<Mod>();
 
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
             {
                 var matchingSelectedMod = SelectedMods.Value.SingleOrDefault(selected => selected.GetType() == modState.Mod.GetType());
 
@@ -495,7 +491,7 @@ private void updateFromInternalSelection()
             if (externalSelectionUpdateInProgress)
                 return;
 
-            var candidateSelection = allAvailableMods.Where(modState => modState.Active.Value)
+            var candidateSelection = AllAvailableMods.Where(modState => modState.Active.Value)
                                                      .Select(modState => modState.Mod)
                                                      .ToArray();
 
diff --git a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
index 56a69be741..294c80677d 100644
--- a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
+++ b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
@@ -97,7 +97,7 @@ protected override void LoadComplete()
         /// </summary>
         private void toggleAllFreeMods()
         {
-            var availableMods = freeModSelectOverlay.AllAvailableAndValidMods.ToArray();
+            var availableMods = allAvailableAndValidMods.ToArray();
 
             Current.Value = Current.Value.Count == availableMods.Length
                 ? Array.Empty<Mod>()
@@ -108,7 +108,7 @@ private void updateModDisplay()
         {
             int current = Current.Value.Count;
 
-            if (current == freeModSelectOverlay.AllAvailableAndValidMods.Count())
+            if (current == allAvailableAndValidMods.Count())
             {
                 count.Text = "all";
                 circle.FadeColour(colours.Yellow, 200, Easing.OutQuint);
@@ -124,5 +124,9 @@ private void updateModDisplay()
                 circle.FadeColour(colours.Gray4, 200, Easing.OutQuint);
             }
         }
+
+        private IEnumerable<Mod> allAvailableAndValidMods => freeModSelectOverlay.AllAvailableMods
+                                                                                 .Where(state => state.ValidForSelection.Value)
+                                                                                 .Select(state => state.Mod);
     }
 }

```

File: osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs [hunk 2]
```diff
@@ -49,14 +87,42 @@ protected override void LoadComplete()
...
+            if (current == freeModSelectOverlay.AllAvailableAndValidMods.Count())
+            {
+                count.Text = "all";
+                circle.FadeColour(colours.Yellow, 200, Easing.OutQuint);
+            }
+            else if (current > 0)
+            {
+                count.Text = $"{current} mods";
+                circle.FadeColour(colours.YellowDark, 200, Easing.OutQuint);
+            }
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs:120]: The foreground-to-background contrast in the "some mods" state is passable, but in the "all" state is pretty dreadful. Propose inverting foreground colour.

```diff
diff --git a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
index 56a69be741..897e7c39c7 100644
--- a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
+++ b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
@@ -111,16 +111,19 @@ private void updateModDisplay()
             if (current == freeModSelectOverlay.AllAvailableAndValidMods.Count())
             {
                 count.Text = "all";
+                count.FadeColour(colours.Gray2, 200, Easing.OutQuint);
                 circle.FadeColour(colours.Yellow, 200, Easing.OutQuint);
             }
             else if (current > 0)
             {
                 count.Text = $"{current} mods";
+                count.FadeColour(colours.Gray2, 200, Easing.OutQuint);
                 circle.FadeColour(colours.YellowDark, 200, Easing.OutQuint);
             }
             else
             {
                 count.Text = "off";
+                count.FadeColour(colours.GrayF, 200, Easing.OutQuint);
                 circle.FadeColour(colours.Gray4, 200, Easing.OutQuint);
             }
         }

```

https://github.com/ppy/osu/assets/20418176/9db1c206-a242-4cbd-a994-c9072f38273c

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs:116]: Aha, I didn't realise that the validity was included in `ModState`. I tried this at first but realised it was selecting invalid mods.

**Comment by peppy** [comment]
> Suggestion: Maybe change "Off" and "On" to "None" and "All"?
> Off/On has the advantage of being very intuitive for new users, especially users coming from stable, but None/All describes better what is actually happening and hints at the possibility of selecting specifc mods. I'm not sure which is better here.

I dunno, I specifically chose "off" and "all" (not what you said here, so maybe you misread?) for ability to understand it.

**Comment by Haspamelodica** [comment]
> I dunno, I specifically chose "off" and "all" (not what you said here, so maybe you misread?) for ability to understand it.

Whoops, yes, I just misread, sorry

**Review by bdach — APPROVED** [review]

**Pull request merged**